src/main/java/core/basesyntax/service/operation/ReturnOperation.java

```
@@ -0,0 +1,18 @@
+package core.basesyntax.service.operation;
+
+import core.basesyntax.model.FruitTransaction;
+import core.basesyntax.service.OperationHandler;
+import core.basesyntax.storage.Storage;
+
+public class ReturnOperation implements OperationHandler {
+    private final Storage storage;
+
+    public ReturnOperation(Storage storage) {
+        this.storage = storage;
+    }
+
+    @Override
+    public void handle(FruitTransaction transaction) {
+        storage.increase(transaction.getFruit(), transaction.getQuantity());
```

The handler directly uses transaction.getFruit() and transaction.getQuantity() without validating them. The task/checklist requires validating input data (fruit name non-null/non-empty and quantity non-negative). Add checks before calling storage.increase, for example:

```java
String fruit = transaction.getFruit();
int qty = transaction.getQuantity();
if (fruit == null || fruit.trim().isEmpty()) {
    throw new FruitShopException("Invalid fruit name in transaction: " + transaction);
}
if (qty < 0) {
    throw new FruitShopException("Negative quantity in transaction: " + transaction);
}
storage.increase(fruit, qty);
```

This ensures the handler fails fast with informative errors rather than allowing invalid state or unclear failures downstream .